### PR TITLE
Fix range slicing mixing uint/int idxTypes

### PR DESCRIPTION
**Pull request by bradcray**
This PR adds the ability to slice (intersect) int ranges with uint ranges and uint ranges with int ranges.  Previously, this was not possible due to the fact that the default min/max operators don't support mixed int/uint types (which is reasonable because of our standard uncertainty about whether the result should be represented as an int/uint).

In the specific context of range slicing, however, we have more context for the operation and can write min() and max() overloads for the mixed int/uint case.  Moreover, these overloads can be written to call out the case when the intersection is empty but not representable in the original range's index set (e.g., if the intersection should include negative indices but the index type is unsigned).  Previously, the spec has said that such cases would result in an error/halt, but in this work it seemed more natural to me to simply define such cases as the empty range 1..0 to avoid halting the program and still get the result of "empty intersection".  (When the indices are representable, it's arguably preferable to continue preserving high/low values so that subsequent range operators might make the empty range into a non-empty one again, preserving the global indices... though in practice, I suspect that such cases will not actually come up much).  Updated the spec to reflect this change in behavior.

Thus, these new min/max overloads return a result of the first argument's index type (and are called in such a way that the slicee range's bounds are always first) and set a variable in the enclosing scope (reset to false on each invocation) to indicate whether the slice should be empty though non-representable.

The previous invocations of these min/max routines are then hoisted out of the range constructor that's going to use their results such that we can capture the values, but then overwrite them to 1..0 if either of the values turned on the "empty intersection" flag.

Retired some futures defined against these cases (updating the one to reflect the empty range rather than the halt/error message) and added in a new test cases that tries a bunch of interesting scenarios (overlapping and non-overlapping int/uint ranges with representable and non-representable indices).

### CODE REVIEW

File: modules/internal/ChapelRange.chpl [hunk 1]
```diff
@@ -1062,12 +1062,83 @@ module ChapelRange {
...
+
+    //
+    // These are mixed int/uint min/max functions that return a value
+    // matching the first argument's type.  They are written knowing
+    // that they will be called in an intersection context and set
+    // this variable if the intersection will be guaranteed to be
+    // empty (e.g., if all values of one range are outside of the
+    // expressible range of the other.
+    //
+    var emptyIntersection: bool;
```
Comment (ronawho) [inline modules/internal/ChapelRange.chpl]: Maybe I'm missing something, but to me it looks like this isn't really used. Was it just for debugging purposes?

File: spec/Ranges.tex [hunk 2]
```diff
@@ -932,7 +932,7 @@ \subsection{Range Slicing}
 \end{rationale}
 
 If the resulting represented sequence cannot be expressed as a range of the
-expected type, an error is generated. This can happen, for example, when
+expected type, the empty range \chpl{1..0} is computed as the result. This can happen, for example, when
```
Comment (ronawho) [inline spec/Ranges.tex]: minor, but this line is pretty long now

File: modules/internal/ChapelRange.chpl (hunk 1 above)
Comment (ronawho) [inline modules/internal/ChapelRange.chpl]: I had a pretty hard time wrapping my head around the use of `emptyIntersection`. Do you think it'd be cleaner to make `emptyIntersection` an out argument or to have min/max return a tuple?

Comment (bradcray) [inline modules/internal/ChapelRange.chpl]: The trouble with doing that is that for any type combination other than int/uint, the calls to max/min in this function dispatch to the ordinary ones (so we can't really change the signatures of them).  These overloads are only used in this difficult case.

File: spec/Ranges.tex (hunk 2 above)
Comment (bradcray) [inline spec/Ranges.tex]: You know that LaTeX ignores whitespace, right?  :)  I was just trying to minimize the diff.

File: modules/internal/ChapelRange.chpl (hunk 1 above)
Comment (ronawho) [inline modules/internal/ChapelRange.chpl]: Ah, got it

**Comment by ronawho** [comment]
good to merge

**Pull request merged**